src/Bicep.Cli/Services/CompilationWriter.cs

```diff
@@ -24,29 +25,60 @@ public EmitResult ToFile(Compilation compilation, string outputPath)
             var fileStream = CreateFileStream(outputPath);
             using (fileStream)
             {
-                var semanticModel = compilation.GetEntrypointSemanticModel();
-                return new TemplateEmitter(semanticModel).Emit(fileStream);
+                return ToStream(compilation, fileStream);           
             }
         }
 
         public EmitResult ToStream(Compilation compilation, Stream stream)
         {
-            var semanticModel = compilation.GetEntrypointSemanticModel();
-            return new TemplateEmitter(semanticModel).Emit(stream);
+            var fileKind = compilation.SourceFileGrouping.EntryPoint.FileKind;
+            switch (fileKind)
+            {
+                case BicepSourceFileKind.BicepFile:
+                    return new TemplateEmitter(compilation.GetEntrypointSemanticModel()).Emit(stream);
+
+                case BicepSourceFileKind.ParamsFile:
+                    return new ParametersEmitter(compilation.GetEntrypointSemanticModel()).EmitParamsFile(stream);
+
+                default:
+                    throw new NotImplementedException($"Unexpected file kind '{fileKind}'");
+            }            
         }
 
         public EmitResult ToStdout(Compilation compilation)
         {
+            var fileKind = compilation.SourceFileGrouping.EntryPoint.FileKind;
             var semanticModel = compilation.GetEntrypointSemanticModel();
-            var sourceFileToTrack = semanticModel.Features.SourceMappingEnabled ? semanticModel.SourceFile : null;
-            using var writer = new SourceAwareJsonTextWriter(semanticModel.FileResolver, invocationContext.OutputWriter, sourceFileToTrack)
+            switch (fileKind)
             {
-                Formatting = Formatting.Indented
-            };
+                case BicepSourceFileKind.BicepFile:
+                    {
+                        var sourceFileToTrack = semanticModel.Features.SourceMappingEnabled ? semanticModel.SourceFile : default;
+                        using var writer = new SourceAwareJsonTextWriter(semanticModel.FileResolver, invocationContext.OutputWriter, sourceFileToTrack)
+                        {
+                            Formatting = Formatting.Indented
+                        };
+
+                        var emitter = new TemplateEmitter(semanticModel);
+
+                        return emitter.Emit(writer);
+                    }
+
+                case BicepSourceFileKind.ParamsFile:
+                    {
+                        using var writer = new JsonTextWriter(invocationContext.OutputWriter)
+                        {
+                            Formatting = Formatting.Indented
+                        };
+
+                        new ParametersEmitter(semanticModel).EmitParamsFile(writer);
 
-            var emitter = new TemplateEmitter(semanticModel);
+                        return new ParametersEmitter(semanticModel).EmitParamsFile(writer);
```

Comment:
Twice for good measure? 😄

Reply:
Guarantees idempotency! ;)

I think this proves we don't have tests for StdOut with params :(